### PR TITLE
feat!: v1

### LINKED ISSUE #1563: http2 upload file from server
hello, its possible create dynamic links for download file?

this is my server

```javascript
const fastify = require('fastify')({
    logger: false,
    http2: true,
    https: {
        key: fs.readFileSync(path.join(serverDirectory, '/certificate/', 'key.pem')),
        cert: fs.readFileSync(path.join(serverDirectory, '/certificate/', 'cert.pem'))
    }
});

fastify.listen(3000, (err, address) => {
    if (err) {
        fastify.log.error(err)
        process.exit(1)
    }
    fastify.log.info(`server listening on ${address}`)
})
```

i need next, pseudo example

```javascript
fastify.get('/download', (req, reply) => {
    searching file used special parametrs on input, and upload a file
    reply.sendFile('file')
})
```

i did not found example...

on express i found example:

Express has a helper for this to make life easier.
```javascript
app.get('/download', function(req, res){
  var file = __dirname + '/upload-folder/dramaticpenguin.MOV';
  res.download(file); // Set disposition and send it.
});
```

### PR DESCRIPTION
**Pull request by metcoder95**
**PR for v1**

This PR introduces v1 implementation for `fastify-download`. Which implements similar feature as `res.download` on [Express.js](https://expressjs.com/es/4x/api.html#res.download) on fastify, decorating `reply` object with a `download` function.

Closes [#1563](https://github.com/fastify/fastify/issues/1563)

**Notes**

- Not sure if is best that the `download` function should return a `Promise`. This is due to the fact that if `lastModified` flag is passed, a promisified version of `fs.stat` it's being called.

**Pending**

- [ ] Testing
- [ ] Documentation
- [ ] Types

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
+const { createReadStream, stat } = require('fs')
```
Comment (mcollina) [inline index.js:2]: use strict is missing

File: index.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    if (Number.isNaN(maxAge) && typeof maxAge !== 'string') {
+      throw new Error("'maxAge' only accepts 'numeric' or 'string' values")
+    }
+
+    return options
+  }
+
+  function validatePath (file, root, allowDotfiles) {
```
Comment (mcollina) [inline index.js:76]: This enables a user to read all files in the filesystem using `..`

**Review by mcollina — CHANGES_REQUESTED** [review]

File: index.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    const fullPath = root ? join(root, path) : path
+    const headers = getHeaders(filename, options)
+
+    if (options.lastModified) {
+      const { mtime } = await statPromise(fullPath)
+
+      this.header('Last-Modified', mtime)
+    }
+
+    const stream = createReadStream(fullPath)
```
Comment (mcollina) [inline index.js:53]: http2 has a `res.respondWithFile` api: https://nodejs.org/dist/latest-v14.x/docs/api/http2.html#http2_http2stream_respondwithfile_path_headers_options

File: index.js (hunk 1 above)
Comment (metcoder95) [reply index.js:2]: I missed that, thanks! Done in 0b997b1

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by mcollina** [comment]
FYI: the original issue was for http2. You might want to consider working on fastify-static instead, which it implements this functionality

File: index.js (hunk 2 above)
Comment (metcoder95) [reply index.js:76]: Got it, I'm not sure if I'm doing it correctly or if I'm fully getting what you mean, but I added a `path.resolve` call here dfa1c14

Basically when trying to read all files using `..`, I'm receiving an `ESDIR` error which is a similar approach to what Express does, see [here](https://github.com/expressjs/express/blob/master/lib/response.js#L435).

Should be done differently?

File: index.js (hunk 3 above)
Comment (metcoder95) [reply index.js:53]: Got it! Missed that of `http2`. Will replace it 👍

**Comment by metcoder95** [comment]
> FYI: the original issue was for http2. You might want to consider working on fastify-static instead, which implements this functionality

Oh that's true. Thanks for pointing me there! I'll open a PR and move the implementation there, taking advantage of the current implementation 👍

**Comment by metcoder95** [comment]
Closed in favor of [#174](https://github.com/fastify/fastify-static/pull/174)

**Pull request closed without merging**